dbcon/joblist/tupleannexconstantstep.cpp

```diff
@@ -0,0 +1,318 @@
+#include "jlf_common.h"
+#include "tupleconstantstep.h"
+#include "limitedorderby.h"
+
+#include "tupleannexconstantstep.h"
+
+#define QUEUE_RESERVE_SIZE 100000
+
+namespace joblist
+{
+TupleAnnexConstantStep::TupleAnnexConstantStep(const JobInfo& jobInfo)
+ : JobStep(jobInfo), fJobList(jobInfo.jobListPtr)
+{
+  fExtendedInfo = "TACS: ";
+  fQtc.stepParms().stepType = StepTeleStats::T_TACS;
+}
+
+TupleAnnexConstantStep::~TupleAnnexConstantStep()
+{
+  if (fConstant)
+    delete fConstant;
+
+  fConstant = NULL;
+}
+
+void TupleAnnexConstantStep::setOutputRowGroup(const rowgroup::RowGroup& rg)
+{
+  throw runtime_error("Disabled, use initialize() to set output RowGroup.");
+}
+
+void TupleAnnexConstantStep::initialize(const RowGroup& rgIn, const JobInfo& jobInfo)
+{
+  // Initialize structures used by separate workers
+  fRowGroupIn = rgIn;
+  fRowGroupIn.initRow(&fRowIn);
+
+  fConstant->initialize(jobInfo, &rgIn);
+  fRowGroupOut = fConstant->getOutputRowGroup();
+
+  fRowGroupOut.initRow(&fRowOut);
+}
+
+void TupleAnnexConstantStep::run()
+{
+  if (fInputJobStepAssociation.outSize() == 0)
+    throw logic_error("No input data list for annex step.");
+
+  fInputDL = fInputJobStepAssociation.outAt(0)->rowGroupDL();
+
+  if (fInputDL == NULL)
+    throw logic_error("Input is not a RowGroup data list.");
+
+  if (fOutputJobStepAssociation.outSize() == 0)
+    throw logic_error("No output data list for annex step.");
+
+  fOutputDL = fOutputJobStepAssociation.outAt(0)->rowGroupDL();
+
+  if (fOutputDL == NULL)
+    throw logic_error("Output is not a RowGroup data list.");
+
+  if (fDelivery == true)
+  {
+    fOutputIterator = fOutputDL->getIterator();
+  }
+
+  fInputDL = fInputJobStepAssociation.outAt(0)->rowGroupDL();
+  if (fInputDL == NULL)
+    throw logic_error("Input is not a RowGroup data list.");
+
+  fInputIterator = fInputDL->getIterator();
+  fRunner = jobstepThreadPool.invoke([this]() { this->execute(); });
+}
+
+void TupleAnnexConstantStep::join()
+{
+  if (fRunner)
+  {
+    jobstepThreadPool.join(fRunner);
+  }
+}
+
+uint32_t TupleAnnexConstantStep::nextBand(messageqcpp::ByteStream& bs)
+{
+  bool more = false;
+  uint32_t rowCount = 0;
+
+  try
+  {
+    bs.restart();
+
+    more = fOutputDL->next(fOutputIterator, &fRgDataOut);
+
+    if (more && !cancelled())
+    {
+      fRowGroupOut.setData(&fRgDataOut);
+      fRowGroupOut.serializeRGData(bs);
+      rowCount = fRowGroupOut.getRowCount();
+    }
+    else
+    {
+      while (more)
+        more = fOutputDL->next(fOutputIterator, &fRgDataOut);
+
+      fEndOfResult = true;
+    }
+  }
+  catch (...)
+  {
+    handleException(std::current_exception(), logging::ERR_IN_DELIVERY, logging::ERR_ALWAYS_CRITICAL,
+                    "TupleAnnexConstantStep::nextBand()");
+    while (more)
+      more = fOutputDL->next(fOutputIterator, &fRgDataOut);
+
+    fEndOfResult = true;
+  }
+
+  if (fEndOfResult)
+  {
+    // send an empty / error band
+    fRgDataOut.reinit(fRowGroupOut, 0);
```

Comment:
Could you pack these 3 calls: setData, resetRowGroup, reset into a RowGroup method and call it appropriately.